qiskit_experiments/database_service/database_service.py

```diff
@@ -0,0 +1,417 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Experiment database service abstract interface."""
+
+from abc import ABC, abstractmethod
+from typing import Optional, Dict, List, Any, Union, Tuple
+
+from qiskit.providers import Options
+
+from .device_component import DeviceComponent
+
+
+class DatabaseService:
+    """Base common type for all versioned DatabaseService abstract classes.
+
+    Note this class should not be inherited from directly, it is intended
+    to be used for type checking. When implementing a subclass you should use
+    the versioned abstract classes as the parent class and not this class
+    directly.
+    """
+
+    version = 0
+
+
+class DatabaseServiceV1(DatabaseService, ABC):
+    """Interface for providing experiment database service.
+
+    This class defines the interface ``qiskit_experiments`` expects from an
+    experiment database service.
+
+    An experiment database service allows you to store experiment data and metadata
+    in a database. An experiment can have one or more jobs, analysis results,
+    and figures.
+
+    Each implementation of this service may use different data structure and
+    should issue a warning on unsupported keywords.
+    """
+
+    version = 1
+
+    def __init__(self):
+        """Initialize an DatabaseService instance."""
+        self._options = self._default_options()
+
+    @classmethod
+    @abstractmethod
+    def _default_options(cls) -> Options:
+        """Return the default options
+
+        Returns:
+            Default options.
+        """
+        pass
+
+    @abstractmethod
+    def create_experiment(
+        self,
+        experiment_type: str,
+        backend_name: str,
+        metadata: Optional[Dict] = None,
+        experiment_id: Optional[str] = None,
+        job_ids: Optional[List[str]] = None,
+        tags: Optional[List[str]] = None,
+        notes: Optional[str] = None,
+        **kwargs: Any,
+    ) -> str:
+        """Create a new experiment in the database.
+
+        Args:
```

Comment:
> It is implemented on the `ibmq-provider` side.

So does qiskit-experiments (this PR I guess) define the interface for working with experiments in Qiskit and then you have different services that implement that interface, and for IBM Quantum that would be the code in qiskit-ibmq-provider? It wasn't clear to me how that's going to work (and I didn't read the large design doc 😄 ).

I was under the impression that the experiment service code in qiskit-ibmq-provider was essentially going to be frozen and be transitioned to this repo.

Reply:
`qiskit-experiments` defines the code to run the actual experiments. It _can_ use IBM Quantum to run circuits and save results, but it doesn't have to. Qiskit is designed to work with different providers (IBM, Honeywell, Aer, etc), and there is no dependency on IBM Quantum. 

This PR essentially defines how it would interact with an "experiment database service", if one is available. IBM Quantum is just the only one providing this service right now.